arch/risc-v/src/common/riscv_percpu.c

```diff
@@ -86,13 +92,39 @@ void riscv_percpu_init(void)
 
 void riscv_percpu_add_hart(uintptr_t hartid)
 {
-  /* Hart IDs go from 0...4 */
+  riscv_percpu_t *percpu;
+  uintptr_t       stack_top;
+
+  /* Get free entry for this hart, this must not fail */
+
+  percpu = (riscv_percpu_t *)sq_remfirst(&g_freelist);
```

Comment:
Because hartid != index. 

For original code that had 5 indices this was true. `CONFIG_ARCH_CPU_COUNT` was needed to allocate a table of 5 entries. But we don't use 5 CPUs.

This decouples hartid entirely from the percpu table, and I think this is a better / more scalable solution.

Reply:
but your code use hartid as index to compute stack location.

Reply:
Hmm yes, that was written based on the old NIRQSTACKS code but you're right, cannot use hartid for that.

We need two things:
- Hartid, the **real** hartid to claim and acknowledge PLIC interrupts
- CPU number, to calculate stack position

One option is to use sl_list for the stacks too, this way the CPU number is not needed.

Reply:
Stack is now assigned to the list, removing dependency to hartid

Reply:
Here is my thought:

1. It's wrong to allocate Remove riscv_percpu from g_freelist without locking
2. Why not change riscv_percpu_add_hart prototype to:
```
void riscv_percpu_init_hart(int index, uintptr_t hartid);
```

Reply:
Answers:
1. Yes, you are right, need lock for sl list. 
2. The it gets hard to define index. Allocating from a free list does not require defining index.

For now the call is done only during init but I'll add the lock to allow hotloading CPUs.

Reply:
Not about hotloading only, cpu1 and cpu2... will boot in parallel.

Reply:
Correct, the lock is needed and now added too

Reply:
_please use flags instead_ I don't know what you mean by flags. Can you give an example ?

I think having the init function and sl list is fine, as it also allows releasing a CPU. There are many other init functions that are called from the chip specific start routines, so I did not think this would be so different from those.

Using, the global variable then leaves the stack unallocated, and a similar mechanism would be needed for that, so it would converge into something like this:

uintptr_t g_init_index;
lock();
riscv_percpu_t *percpu = &g_percpu[g_init_index++];
unlock();

That implementation is something I tried to avoid by using sl list.

Reply:
> _please use flags instead_ I don't know what you mean by flags. Can you give an example ?
> 
> I think having the init function and sl list is fine, as it also allows releasing a CPU. There are many other init functions that are called from the chip specific start routines, so I did not think this would be so different from those. And yes, `__mpfs_start()` must be fixed to work with SMP. All other init routines are also executed on all harts too, so currently with SMP calling `__mpfs_start()` will fail miserably.
> 
> Using, the global variable then leaves the stack unallocated, and a similar mechanism would be needed for that, so it would converge into something like this:
> 
> uintptr_t g_init_index; lock(); riscv_percpu_t *percpu = &g_percpu[g_init_index++]; unlock();
> 
> That implementation is something I tried to avoid by using sl list.

Even using sl list, you can do the initialization when riscv_percpu_add_hart is called first time to simplify the use.

Reply:
Done

Reply:
Hmm, I don't think the spinlock does quite what I intended. Need mutex or semaphore but those are not available yet.